x/gamm/twap/logic_test.go

```diff
@@ -0,0 +1,206 @@
+			require.Equal(t, test.expRecord, gotRecord)
+		})
+	}
+}
+
+func (s *TestSuite) TestUpdateTwap() {
+	poolId := s.PrepareUni2PoolWithAssets(defaultUniV2Coins[0], defaultUniV2Coins[1])
+	newSp := sdk.OneDec()
+
+	tests := map[string]struct {
+		record     types.TwapRecord
+		updateTime time.Time
+		expRecord  types.TwapRecord
+	}{
+		"0 accum start": {
+			record:     newRecord(time.Unix(1, 0), sdk.NewDec(10), zeroDec, zeroDec),
+			updateTime: time.Unix(2, 0),
+			expRecord:  newExpRecord(OneSec.MulInt64(10), OneSec.QuoInt64(10)),
+		},
+	}
+	for name, test := range tests {
+		s.Run(name, func() {
+			// setup common, block time, pool Id, expected spot prices
+			s.Ctx = s.Ctx.WithBlockTime(test.updateTime.UTC())
+			test.record.PoolId = poolId
+			test.expRecord.PoolId = poolId
+			test.expRecord.P0LastSpotPrice = newSp
+			test.expRecord.P1LastSpotPrice = newSp
+			test.expRecord.Height = s.Ctx.BlockHeight()
+			test.expRecord.Time = s.Ctx.BlockTime()
+
+			newRecord := s.twapkeeper.UpdateRecord(s.Ctx, test.record)
+			s.Require().Equal(test.expRecord, newRecord)
+		})
+	}
+}
+
+// TestComputeArithmeticTwap tests ComputeArithmeticTwap on various inputs.
+// The test vectors are structured by setting up different start and records,
+// based on time interval, and their accumulator values.
+// Then an expected TWAP is provided in each test case, to compare against computed.
+func TestComputeArithmeticTwap(t *testing.T) {
+	denom0, denom1 := "token/B", "token/A"
+	newOneSidedRecord := func(time time.Time, accum sdk.Dec, useP0 bool) types.TwapRecord {
+		record := types.TwapRecord{Time: time, Asset0Denom: denom0, Asset1Denom: denom1}
+		if useP0 {
+			record.P0ArithmeticTwapAccumulator = accum
+		} else {
+			record.P1ArithmeticTwapAccumulator = accum
+		}
+		record.P0LastSpotPrice = sdk.ZeroDec()
+		record.P1LastSpotPrice = sdk.OneDec()
+		return record
+	}
+
+	type testCase struct {
+		startRecord types.TwapRecord
+		endRecord   types.TwapRecord
+		quoteAsset  string
+		expTwap     sdk.Dec
+	}
+
+	testCaseFromDeltas := func(startAccum, accumDiff sdk.Dec, timeDelta time.Duration, expectedTwap sdk.Dec) testCase {
+		return testCase{
+			newOneSidedRecord(baseTime, startAccum, true),
+			newOneSidedRecord(baseTime.Add(timeDelta), startAccum.Add(accumDiff), true),
+			denom0,
+			expectedTwap,
+		}
+	}
+	testCaseFromDeltasAsset1 := func(startAccum, accumDiff sdk.Dec, timeDelta time.Duration, expectedTwap sdk.Dec) testCase {
+		return testCase{
+			newOneSidedRecord(baseTime, startAccum, false),
+			newOneSidedRecord(baseTime.Add(timeDelta), startAccum.Add(accumDiff), false),
+			denom1,
+			expectedTwap,
+		}
+	}
+	tenSecAccum := OneSec.MulInt64(10)
+	pointOneAccum := OneSec.QuoInt64(10)
+	plusOneSec := baseTime.Add(time.Second)
+	tests := map[string]testCase{
+		"basic: spot price = 1 for one second, 0 init accumulator": {
+			newOneSidedRecord(baseTime, sdk.ZeroDec(), true),
+			newOneSidedRecord(plusOneSec, OneSec, true),
```

Comment:
Why are we providing `OneSec` for the accumulator?

Reply:
I understand why now. Just got confused initially because the accumulator is the sum of the product of time and price